### PR TITLE
feat(ui): Group nodes based on `templateRef` and show invoking template name. Fixes #11106

### LINKED ISSUE #11106: Group UI nodes based on `templateRef`
# Summary

I was chatting with @mshatkhin23, and he mentioned that he'd like to have the workflow graph grouped based on the workflow template ref. Given this workflow snippet (using templates instead of refs here for simplicity):
```yaml
- name: dag-tier-3
  dag:
    tasks:
      - name: task-1
        template: flakey-container
      - name: task-2
        depends: 'task-1'
        template: flakey-container
- name: dag-tier-2
  dag:
    tasks:
      - name: task-1
        template: dag-tier-3
      - name: task-2
        depends: 'task-1'
        template: dag-tier-3
- name: user-workflow
  dag:
    tasks:
      - name: task-1
        template: dag-tier-2
      - name: task-2
        depends: 'task-1'
        template: dag-tier-2
```
It produces the following graph:
image

We'd like to find a way to group or collapse things based on a given template. In the graph above, it's hard to know which `task-1` is in user-workflow or `dag-tier-2` or `dag-tier-3`. Ideally we:
- collapse templateRefs from the start
- adjust the display name so that the task name also includes which template is invoking it

# Use Cases

This should help eliminate confusion when looking at the Argo Workflows graph. As workflows scale and orgs manage lots of templates, there will naturally be naming collisions that arise

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by Adrien-D**
Fixes #11106.

### Motivation

It's confusing to look at the Workflows graph with templateRef and tasks having the same names. Creates linear graph without template hierarchy.

### Modifications

- Added the ability to show the invoking template name on nodes. To keep the actual display, I added the switch button on the toolbar. Also updated the title to include the full id and name of the node as the test is truncated.
- Added the grouping by templateRef using `boundaryId`. This seems to have the desired results. 

Demo

https://github.com/user-attachments/assets/763ea3b5-7349-405f-8fa9-420d9a8a091e

### Verification

I tested on simple workflows, especially the one listed on the original issue, see screen recording

### CODE REVIEW

**Comment by mshatkhin23** [comment]
@Adrien-D thanks for working on this! looks great

Clarifying question - do I understand correctly that collapsing is only happening for templateRefs (I.e. templates from other WorkflowTemplates)? 

If we have a task/step that references another template defined in our Workflow / Workflow Template, would that also be collapsible?

**Comment by Adrien-D** [comment]
@mshatkhin23 yes collapsing will only happen on `templateRefs`

> If we have a task/step that references another template defined in our Workflow / Workflow Template, would that also be collapsible?

How do you reference it ? Can you give a very small example ?

**Comment by djanjic** [comment]
Tested locally multiple use-cases and everything went well.

**Comment by Adrien-D** [comment]
@mshatkhin23 It will only be collapsible by `templateRefs`, so 
in this example that won't collapse 
```
- name: user-workflow
  dag:
    tasks:
      - name: task-1
        template: dag-tier-2
      - name: task-2
        depends: 'task-1'
        template: dag-tier-2
```

whereas is this case it will collapse
```
- name: user-workflow
  dag:
    tasks:
      - name: task-1
        templateRef:
          clusterScope: true/false
          name: external-workflow-template
          template: external-dag-tier-2
          ```

**Comment by Adrien-D** [comment]
Happy to merge this one as it is and discuss about collapsing on `templates` in next issue !

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx [hunk 1]
```diff
@@ -100,8 +92,28 @@ export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRe
         );
     }
 
+    private nodeLabel(n: NodeStatus, parent?: NodeStatus) {
+        const phase = n.type === 'Suspend' && n.phase === 'Running' ? 'Suspended' : n.phase;
+        let label = shortNodeName(n);
+
+        if (this.state.showInvokingTemplateName) {
+            label = `${parent?.templateRef?.name ?? parent?.templateName ?? n.templateName}:${label}`;
+        }
```
Comment (isubasinghe) [inline ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:101]: I thought this would break for inline templates but it works, nice work!

**Review by isubasinghe — APPROVED** [review]
LGTM -> I tested this out and worked.

This PR has however raised some questions about the logic used in the traverse function. (You didn't author this).

**Comment by isubasinghe** [comment]
TODO: Investigate why `allNodes` in workflow-dag.tsx is indexed by nodeID but then later on indexed by nodeName.

**Pull request merged**